src/blob.rs

```diff
@@ -1,86 +1,258 @@
+#[derive(Debug)]
+pub struct WriteFuture {
+    file: fs::File,
+    buf: Option<Vec<u8>>,
+    current_offset: u64,
 }
 
-impl<K> Blob<K>
-where
-    K: Default,
-{
-    /// # Description
-    /// Create new blob from file
-    // @TODO more useful result
-    pub fn from_file<P: AsRef<Path>>(_path: P) -> Result<Self, ()> {
-        // @TODO implement
-        Ok(Self::default())
+impl Future for WriteFuture {
+    type Output = Result<()>;
+
+    fn poll(mut self: Pin<&mut Self>, _waker: &Waker) -> Poll<Self::Output> {
+        let buf = self.buf.take().unwrap();
+        self.file.write_at(&buf, self.current_offset).unwrap();
+        Poll::Ready(Ok(()))
     }
+}
 
-    /// # Description
-    /// Writes given slice to file
-    // @TODO more useful result
-    pub fn write(&mut self, _record: Record<K>) -> Result<(), ()> {
-        // @TODO implement
-        Ok(())
+#[derive(Debug)]
+pub struct ReadFuture {
+    file: fs::File,
+    key: Vec<u8>,
+    loc: Location,
+}
+
+impl Future for ReadFuture {
+    type Output = Result<Record>;
+
+    fn poll(self: Pin<&mut Self>, _waker: &Waker) -> Poll<Self::Output> {
+        let mut buf = vec![0u8; self.loc.size as usize];
+        self.file.read_at(&mut buf, self.loc.offset).unwrap();
+        Poll::Ready(Record::from_raw(&buf).map_err(Error::RecordError))
     }
+}
 
+impl Blob {
     /// # Description
-    /// Reads record data, yields `Ok(Vec<u8>)` if read successful,
-    /// otherwise - `Err`
-    // @TODO more useful result
-    pub fn read(&self, _key: &K) -> Result<Record<K>, ()> {
-        // @TODO implement
-        Ok(Record::new())
+    /// Creates new blob file
+    pub fn open_new(name: FileName) -> Result<Self> {
+        Ok(Self {
+            header: Header::new(),
+            file: fs::OpenOptions::new()
+                .create_new(true)
+                .append(true)
+                .read(true)
+                .open(name.as_path())
+                .map_err(Error::OpenNew)?,
+            name,
+            index: BTreeMap::new(),
+            current_offset: 0,
+        })
+    }
+
+    pub fn boxed(self) -> Box<Self> {
+        Box::new(self)
+    }
+
+    pub fn from_file(path: PathBuf) -> Result<Self> {
+        let file = fs::OpenOptions::new()
+            .create(false)
+            .append(true)
+            .read(true)
+            .open(&path)
+            .map_err(Error::FromFile)?;
+        let name = FileName::from_path(&path)?;
+        let len = file.metadata().map_err(Error::FromFile)?.len();
+        // @TODO Scan existing file to create index
+        Ok(Self {
+            header: Header::new(),
+            file,
+            name,
+            index: Default::default(),
+            current_offset: len,
+        })
+    }
+
+    pub fn write(&mut self, record: &mut Record) -> Result<WriteFuture> {
+        record.set_blob_offset(self.current_offset);
+        self.index
+            .insert(record.key().to_vec(), record.header().clone());
```

Comment:
What happen if the data with the same key already exists in this BLOB?

Reply:
should fail now

Reply:
Return ```Error::AlreadyContainsSameKey```